### ISSUE TITLE
[Solution] Reason for high fee estimates from insight.bitpay.com

### ISSUE THREAD

**Issue by dabura667** [bitpay/copay#7203]
I was wondering why insight.bitpay.com and trezor's insight give different values for estimate fees.

It's because your node doesn't see segwit, so it bases the fee/byte calculations based on ONLY non-segwit bytes. That way segwit transactions out on the network look like they are paying higher sat/byte than they actually are.

Judging from the difference between trezor's insight (which all they did was rebase the patches onto >0.14) you could save your users 38% fees just by pointing your public BWS to trezor's insight instead. (If you ask them nicely, I'm sure they wouldn't mind)

https://insight.bitpay.com/api/utils/estimatefee?nbBlocks=2    ==  {"2":0.00697425}
https://btc-bitcore1.trezor.io/api/utils/estimatefee?nbBlocks=2   ==   {"2":0.00430373}

(for non-technical people reading this: This is not "support segwit to save fees" at all. The problem here is that their bitcore fee estimation algorithm is basing the fee estimation on a FEE / BYTES of transactions in the previous blocks, but when it sees a Segwit transaction, it ignores the witness completely, so BYTES is much smaller than actual, and the fee rate calculated of the confirmed transactions is 25-40% higher than the ACTUAL bitcoin core estimation (which is still horrible, estimatefee is deprecated for estimatesmartfee)...)

So, some simple fixes to make a better user experience:

1. Migrate the public insight instance to a segwit-validating version of Bitcoin software. Or just ask Trezor to use their instance for fee estimations for the time being.
2. Migrate the insight fee estimation to estimatesmartfee.
3. Migrate bitcore-wallet-client to using segwit.

Out of the three, the one with the biggest impact would be 2... but 2 pretty much requires 1, and 1 will already cut estimates almost in half. 3 is a great move but tbh a flat 40-50% byte size saving from segwit is not as significant as getting fee estimates right, since fee estimates from bitpay's insight can be unreliable and jump 2-3x randomly without an actual effect on the mempool, basically negating any savings segwit would provide.

**Comment by Josema** [comment]
Do you know any other API that gives a more accurate fees? I saw this https://bitcoinfees.21.co/api/v1/fees/recommended but I don't know how to calculate it. The value that it gives is per byte. But I don't know how many bytes have my transaction. Do you know if the Hexadecimal rawtx string are the bytes of the transaction?

`fee = stringRawTxHex.length * response. fastestFee`

Is this correct?

**Comment by junderw** [comment]
@Josema 
Yeah... comparing trezor and bitpay, trezor (still using the same estimatefee, just with the witness data ADDED (discounted) IN THE CALCULATION instead of NOT AT ALL) Trezor is 45% cheaper across the board for every nBlocks... hooray segwit adoption... but as segwit adoption increases bitpay's fee estimates will get higher and higher until it will estimate 1000 sat/byte during a time of 1 sat/byte confirming.

https://www.bitgo.com/api/v1/tx/fee?numBlocks=2

bitgo is a good fee estimator... slightly closer to reality than trezor... but for some reason it screws up estimates >= nBlocks 10... so be careful. (look at it below)

`{"feePerKb":116898,"cpfpFeePerKb":116898,"numBlocks":9,"confidence":80,"multiplier":1,"feeByBlockTarget":{"1":170675,"2":157345,"3":149393,"5":141465,"6":121631,"9":116898,"10":121679,"11":116658,"15":95869,"17":91196,"41":87096,"42":116557,"49":105800,"73":61425,"97":59252,"121":56030,"145":16577,"313":15028,"337":13545,"361":13015,"505":12269,"721":11120,"745":5597,"817":5392,"865":5083}}`

10 is higher than 9 lol... but the <10 estimates have always been very accurate in my experience. (Our backend uses Bitgo)

**Comment by junderw** [comment]
@Josema 

`fee = stringRawTxHex.length / 2 * response.fastestFee`

you need to divide by two... if you have a hex string of "01ff" its length is 4, but it is representing 2 bytes of data. Divide the hex string by two.
